File: publisher.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"time"
	"flag"


	MQTT "github.com/eclipse/paho.mqtt.golang"
)


type Configuration struct {
	Unit             string  `json:"unit"`
	TransmissionRate float64 `json:"transmission_rate_hz"`
	Longitude        float64 `json:"longitude"`
	Latitude         float64 `json:"latitude"`
	Sensor           string  `json:"sensor"`
	QoS			     byte    `json:"qos"`
}

type Data struct {
	Value            float64   `json:"value"`
	Unit             string    `json:"unit"`
	TransmissionRate float64   `json:"transmission_rate"`
	Longitude        float64   `json:"longitude"`
	Latitude         float64   `json:"latitude"`
	Sensor           string    `json:"sensor"`
	Timestamp        time.Time `json:"timestamp"`
	QoS			  byte      `json:"qos"`
}

type MQTTConnector interface {
	Connect(nodeName string, username string, password string) MQTT.Client
}

type LocalMQTTConnector struct{}

func (l *LocalMQTTConnector) Connect(nodeName string, username string, password string) MQTT.Client {
	opts := MQTT.NewClientOptions().AddBroker("tcp://localhost:1891")
	opts.SetClientID(nodeName)
	client := MQTT.NewClient(opts)

	if token := client.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	return client
}

type HiveMQConnector struct{}

func (h *HiveMQConnector) Connect(nodeName string, username string, password string) MQTT.Client {
	opts := MQTT.NewClientOptions().AddBroker("tls://b9f3c31144f64d469f184727678d8fb6.s1.eu.hivemq.cloud:8883/mqtt")
	opts.SetClientID(nodeName)
	opts.SetUsername(username)
	opts.SetPassword(password)
	client := MQTT.NewClient(opts)

	if token := client.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	return client
}

func main() {
	configPath := flag.String("config", "", "Path to the configuration file")
	csvPath := flag.String("csv", "", "Path to the CSV file")
	connection := flag.String("connection", "hivemq", "Enter 'hivemq' or 'local' for MQTT connection")

	hivemqUsername := flag.String("username", "", "HiveMQ username")
	hivemqPassword := flag.String("password", "", "HiveMQ password")

	flag.Parse()

	if *configPath == "" || *csvPath == "" {
		fmt.Println("Usage: go run publisher.go -config <config_path> -csv <csv_path> -connection <hivemq/local>")
		return
	}

	config, err := readConfig(*configPath)
	if err != nil {
		panic(err)
	}

	var connector MQTTConnector

	if *connection == "hivemq"{
		connector = &HiveMQConnector{}
	} else if *connection == "local" {
		connector = &LocalMQTTConnector{}
	} else {
		fmt.Println("Invalid connection type. Enter 'hivemq' or 'local'")
		return	
	}

	client := connector.Connect("publisher", *hivemqUsername, *hivemqPassword)
	defer client.Disconnect(250)

	data, err := readCSV(*csvPath)
	if err != nil {
		panic(err)
	}

	publishData(client, config, data)
}

func readCSV(csvPath string) ([]float64, error) {
	file, err := os.Open(csvPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var values []float64
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()

		var value float64
		_, err := fmt.Sscanf(line, "%f", &value)
		if err != nil {
			return nil, err
		}

		values = append(values, value)
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return values, nil
}

func publishData(client MQTT.Client, config Configuration, data []float64) {
	interval := time.Second / time.Duration(config.TransmissionRate)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for _, value := range data {
		roundedValue := math.Round(value*100) / 100

		fmt.Printf("Publishing: %f\n", roundedValue)

		message := createJSONMessage(config, roundedValue)

		token := client.Publish("sensor/"+config.Sensor, byte(config.QoS), false, message)
		token.Wait()

		<-ticker.C
	}
}

func createJSONMessage(config Configuration, roundedValue float64) []byte {
	data := Data{
		Value:            roundedValue,
		Unit:             config.Unit,
		TransmissionRate: config.TransmissionRate,
		Longitude:        config.Longitude,
		Latitude:         config.Latitude,
		Sensor:           config.Sensor,
		Timestamp:        time.Now(),
		QoS:			  config.QoS,
	}

	jsonMsg, err := json.Marshal(data)
	if err != nil {
		panic(err)
	}

	return jsonMsg
}

func readConfig(filename string) (Configuration, error) {
	file, err := os.Open(filename)
	if err != nil {
		return Configuration{}, err
	}
	defer file.Close()

	decoder := json.NewDecoder(file)
	config := Configuration{}
	err = decoder.Decode(&config)
	return config, err
}

File: ponderada/mqtt/publisher.py
import json
import csv
import math
import argparse
from datetime import datetime, timedelta
import paho.mqtt.client as mqtt
import time
import paho.mqtt.client as paho
from paho import mqtt


class Configuration:
    def __init__(self, unit, transmission_rate_hz, longitude, latitude, sensor, qos):
        self.unit = unit
        self.transmission_rate_hz = transmission_rate_hz
        self.longitude = longitude
        self.latitude = latitude
        self.sensor = sensor
        self.qos = qos

class Data:
    def __init__(self, value, unit, transmission_rate, longitude, latitude, sensor, timestamp, qos):
        self.value = value
        self.unit = unit
        self.transmission_rate = transmission_rate
        self.longitude = longitude
        self.latitude = latitude
        self.sensor = sensor
        self.timestamp = timestamp
        self.qos = qos


    
def on_connect(client, userdata, flags, reason_code, properties):
    print(f"CONNACK received with code {reason_code}")


def read_config(filename):
    with open(filename, 'r') as f:
        data = json.load(f)
        return Configuration(**data)

def read_csv(csv_path):
    values = []
    with open(csv_path, newline='') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            values.append(float(row[0]))
    return values

def publish_data(client, config, data):
    interval = timedelta(seconds=1/config.transmission_rate_hz)
    for value in data:
        rounded_value = round(value, 2)
        print(f"Publishing: {rounded_value}")
        message = create_json_message(config, rounded_value)
        client.publish(f"sensor/{config.sensor}", message, qos=config.qos)
        time.sleep(interval.total_seconds())

def create_json_message(config, rounded_value):
    data = Data(
        value=rounded_value,
        unit=config.unit,
        transmission_rate=config.transmission_rate_hz,
        longitude=config.longitude,
        latitude=config.latitude,
        sensor=config.sensor,
        timestamp=datetime.now().isoformat(),
        qos=config.qos
    )
    return json.dumps(data.__dict__)

def main():
    broker_address = 'b9f3c31144f64d469f184727678d8fb6.s1.eu.hivemq.cloud'
    port = 8883
    config = read_config('config.json')
    values = read_csv('data.csv')
    topic = f"sensor/{config.sensor}"
    username = 'elisa'
    password = 'Elisa123'

    client = paho.Client(paho.CallbackAPIVersion.VERSION2, "Publisher",
                     protocol=paho.MQTTv5)

    client.on_connect = on_connect

    # Configurações de TLS
    client.tls_set(tls_version=mqtt.client.ssl.PROTOCOL_TLS)
    client.username_pw_set(username, password)  # Configuração da autenticação

    # Conexão ao broker
    client.connect(broker_address, port=port)

    publish_data(client, config, values)

if __name__ == "__main__":
    main()
